lib/statsd/instrument/batched_udp_sink.rb

```diff
@@ -26,112 +15,132 @@ def self.for_addr(addr, flush_interval: DEFAULT_FLUSH_INTERVAL)
-          spawn_dispatcher
-        end
-
-        @buffer << datagram
-        self
+        @dispatcher << datagram
       end
 
-      def shutdown
-        @interrupted = true
-        @dispatcher_thread&.join
-        invalidate_socket
-      end
+      class Dispatcher
+        BUFFER_CLASS = if !::Object.const_defined?(:RUBY_ENGINE) || RUBY_ENGINE == "ruby"
+          ::Array
+        else
+          begin
+            gem("concurrent-ruby")
+          rescue Gem::MissingSpecError
+            raise Gem::MissingSpecError, "statsd-instrument depends on `concurrent-ruby` on #{RUBY_ENGINE}"
+          end
+          require "concurrent/array"
+          Concurrent::Array
+        end
 
-      private
+        def initialize(host, port, flush_interval)
+          @host = host
+          @port = port
+          @interrupted = false
+          @flush_interval = flush_interval
+          @buffer = BUFFER_CLASS.new
+          @dispatcher_thread = Thread.new { dispatch }
+        end
 
-      def spawn_dispatcher
-        @dispatcher_thread = Thread.new { dispatch }
-      end
+        def <<(datagram)
+          unless @dispatcher_thread&.alive?
+            @buffer.clear
+            @dispatcher_thread = Thread.new { dispatch }
+          end
+        end
 
-      NEWLINE = "\n".b.freeze
-      def flush
-        return if @buffer.empty?
+        def shutdown
+          @interrupted = true
+        end
 
-        datagrams = @buffer.shift(@buffer.size)
+        private
 
-        until datagrams.empty?
-          packet = String.new(datagrams.pop, encoding: Encoding::BINARY, capacity: MAX_PACKET_SIZE)
+        NEWLINE = "\n".b.freeze
+        def flush
+          return if @buffer.empty?
 
-          until datagrams.empty? || packet.bytesize + datagrams.first.bytesize + 1 > MAX_PACKET_SIZE
-            packet << NEWLINE << datagrams.shift
-          end
+          datagrams = @buffer.shift(@buffer.size)
 
-          send_packet(packet)
-        end
-      end
+          until datagrams.empty?
+            packet = String.new(datagrams.pop, encoding: Encoding::BINARY, capacity: MAX_PACKET_SIZE)
 
-      def dispatch
-        until @interrupted
-          begin
-            start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
-            flush
-            next_sleep_duration = @flush_interval - (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start)
+            until datagrams.empty? || packet.bytesize + datagrams.first.bytesize + 1 > MAX_PACKET_SIZE
+              packet << NEWLINE << datagrams.shift
+            end
 
-            sleep(next_sleep_duration) if next_sleep_duration > 0
-          rescue => error
-            report_error(error)
+            send_packet(packet)
           end
         end
-      end
 
-      def report_error(error)
-        StatsD.logger.error do
-          "[#{self.class.name}] The dispatcher thread encountered an error #{error.class}: #{error.message}"
+        def dispatch
+          until @interrupted
```

Comment:
Did you want to change this to `begin/until` ?

Reply:
No, I just call `flush` on exit.